File: internal/overlay/presentation_test.go
package overlay

import (
	"encoding/json"
	"math"
	"strings"
	"testing"
	"time"
)

func TestPresentationMarshalShapeAndConstructors(t *testing.T) {
	observedAt := time.Date(2026, 7, 16, 12, 0, 0, 0, time.UTC)
	progress := 0.5
	presentation := Presentation{
		Schema:       PresentationSchemaV1,
		GameID:       "palworld",
		UserID:       "steam_1",
		ObservedAt:   observedAt,
		FreshUntil:   observedAt.Add(15 * time.Second),
		SourceStatus: "online",
		Identity:     Identity{DisplayName: "Player"},
		Map:          &MapPosition{X: 1, Y: 2, Projection: "palworld_world_v1", TileSet: "palworld_default_v1", TileURL: "/map/tiles/{z}/{x}/{y}.png"},
		Fields: []DisplayField{
			StringDisplayField("identity.account", "Account", "text", "account_1", "normal", nil),
			NumberDisplayField("network.latency", "Latency", "latency_ms", 38.5, "normal", &progress),
			TimestampDisplayField("presence.last_online", "Last online", observedAt, "muted"),
			CoordinatesDisplayField("location.coordinates", "Coordinates", 1, 2, "normal"),
			UnavailableDisplayField("policy.remaining", "Remaining", "duration_ms", "muted"),
		},
	}
	if err := ValidatePresentation(presentation); err != nil {
		t.Fatalf("ValidatePresentation() error = %v", err)
	}

	data, err := json.Marshal(presentation)
	if err != nil {
		t.Fatalf("marshal presentation: %v", err)
	}
	root := decodeJSONObject(t, data)
	assertExactJSONKeys(t, root, "schema", "game_id", "user_id", "observed_at", "fresh_until", "source_status", "identity", "map", "fields")
	fields := decodeJSONRawArray(t, root["fields"])
	assertExactJSONKeys(t, decodeJSONObject(t, fields[0]), "id", "label", "kind", "available", "value", "tone")
	assertExactJSONKeys(t, decodeJSONObject(t, fields[1]), "id", "label", "kind", "available", "value", "tone", "progress")
	assertExactJSONKeys(t, decodeJSONObject(t, fields[4]), "id", "label", "kind", "available", "tone")
}

func TestDisplayFieldConstructorsRejectMismatchedKindsAndNonFiniteValues(t *testing.T) {
	tests := []struct {
		name  string
		build func()
	}{
		{name: "string with number kind", build: func() { StringDisplayField("field.one", "One", "integer", "1", "normal", nil) }},
		{name: "number with string kind", build: func() { NumberDisplayField("field.one", "One", "text", 1, "normal", nil) }},
		{name: "non-finite number", build: func() { NumberDisplayField("field.one", "One", "latency_ms", math.NaN(), "normal", nil) }},
		{name: "non-finite coordinate", build: func() { CoordinatesDisplayField("field.one", "One", math.Inf(1), 2, "normal") }},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			defer func() {
				if recover() == nil {
					t.Fatal("constructor did not reject invalid input")
				}
			}()
			tt.build()
		})
	}
}

func TestValidatePresentationRejectsInvalidFields(t *testing.T) {
	valid := StringDisplayField("identity.account", "Account", "text", "account", "normal", nil)
	progress := 0.5
	tests := []struct {
		name   string
		fields []DisplayField
		want   string
	}{
		{name: "duplicate ID", fields: []DisplayField{valid, valid}, want: "duplicate"},
		{name: "unsafe ID", fields: []DisplayField{{ID: "identity/IP", Label: "IP", Kind: "text", Available: true, Value: json.RawMessage(`"secret"`), Tone: "normal"}}, want: "unsafe"},
		{name: "unsupported kind", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "html", Available: true, Value: json.RawMessage(`"x"`), Tone: "normal"}}, want: "kind"},
		{name: "unsupported tone", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "text", Available: true, Value: json.RawMessage(`"x"`), Tone: "critical"}}, want: "tone"},
		{name: "available missing value", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "text", Available: true, Tone: "normal"}}, want: "value"},
		{name: "unavailable carries value", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "text", Value: json.RawMessage(`"x"`), Tone: "muted"}}, want: "unavailable"},
		{name: "unavailable carries progress", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "duration_ms", Tone: "muted", Progress: &progress}}, want: "unavailable"},
		{name: "progress below zero", fields: []DisplayField{withProgress(valid, -0.01)}, want: "progress"},
		{name: "progress above one", fields: []DisplayField{withProgress(valid, 1.01)}, want: "progress"},
		{name: "progress NaN", fields: []DisplayField{withProgress(valid, math.NaN())}, want: "progress"},
		{name: "text wrong JSON type", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "text", Available: true, Value: json.RawMessage(`1`), Tone: "normal"}}, want: "text"},
		{name: "status malformed JSON", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "status", Available: true, Value: json.RawMessage(`"online`), Tone: "normal"}}, want: "status"},
		{name: "integer fractional", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "integer", Available: true, Value: json.RawMessage(`1.5`), Tone: "normal"}}, want: "integer"},
		{name: "integer precision-hidden fraction", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "integer", Available: true, Value: json.RawMessage(`9007199254740992.1`), Tone: "normal"}}, want: "integer"},
		{name: "duration string", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "duration_ms", Available: true, Value: json.RawMessage(`"1000"`), Tone: "normal"}}, want: "duration_ms"},
		{name: "latency object", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "latency_ms", Available: true, Value: json.RawMessage(`{}`), Tone: "normal"}}, want: "latency_ms"},
		{name: "timestamp invalid", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "timestamp", Available: true, Value: json.RawMessage(`"yesterday"`), Tone: "normal"}}, want: "timestamp"},
		{name: "coordinates missing y", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "coordinates", Available: true, Value: json.RawMessage(`{"x":1}`), Tone: "normal"}}, want: "coordinates"},
		{name: "coordinates extra data", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "coordinates", Available: true, Value: json.RawMessage(`{"x":1,"y":2,"ip":"secret"}`), Tone: "normal"}}, want: "coordinates"},
		{name: "coordinates duplicate x", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "coordinates", Available: true, Value: json.RawMessage(`{"x":"hidden","x":1,"y":2}`), Tone: "normal"}}, want: "coordinates"},
		{name: "coordinates duplicate y", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "coordinates", Available: true, Value: json.RawMessage(`{"x":1,"y":0,"y":2}`), Tone: "normal"}}, want: "coordinates"},
		{name: "coordinates trailing JSON", fields: []DisplayField{{ID: "field.one", Label: "One", Kind: "coordinates", Available: true, Value: json.RawMessage(`{"x":1,"y":2} true`), Tone: "normal"}}, want: "coordinates"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			err := ValidatePresentation(Presentation{Schema: PresentationSchemaV1, Fields: tt.fields})
			if err == nil || !strings.Contains(strings.ToLower(err.Error()), tt.want) {
				t.Fatalf("ValidatePresentation() error = %v, want containing %q", err, tt.want)
			}
		})
	}
}

func TestValidatePresentationAcceptsExactIntegerAndCoordinateBoundaries(t *testing.T) {
	tests := []DisplayField{
		{ID: "integer.large", Label: "Large", Kind: "integer", Available: true, Value: json.RawMessage(`9007199254740992`), Tone: "normal"},
		{ID: "integer.exponent", Label: "Exponent", Kind: "integer", Available: true, Value: json.RawMessage(`1e3`), Tone: "normal"},
		{ID: "coordinates.reordered", Label: "Coordinates", Kind: "coordinates", Available: true, Value: json.RawMessage(`{"y":-2.5e2,"x":1.25}`), Tone: "normal"},
	}
	if err := ValidatePresentation(Presentation{Schema: PresentationSchemaV1, Fields: tests}); err != nil {
		t.Fatalf("ValidatePresentation() error = %v", err)
	}
}

func withProgress(field DisplayField, progress float64) DisplayField {
	field.Progress = &progress
	return field
}

func decodeJSONRawArray(t *testing.T, data []byte) []json.RawMessage {
	t.Helper()
	var values []json.RawMessage
	if err := json.Unmarshal(data, &values); err != nil {
		t.Fatalf("decode JSON array: %v", err)
	}
	return values
}

File: overlay/src/contracts/presentation.test.ts
// @vitest-environment node

import { describe, expect, expectTypeOf, it } from 'vitest'

import {
  parsePresentation,
  type AvailableDisplayField,
  type DisplayField,
} from './presentation'

function field(overrides: Record<string, unknown> = {}): Record<string, unknown> {
  return {
    id: 'identity.account',
    label: '账号',
    kind: 'text',
    available: true,
    value: 'Lamball Keeper',
    tone: 'normal',
    ...overrides,
  }
}

function presentation(
  fields: Record<string, unknown>[] = [field()],
): Record<string, unknown> {
  return {
    schema: 'overlay.presentation/v1',
    game_id: 'palworld',
    user_id: 'steam_player',
    observed_at: '2026-07-17T08:00:00Z',
    fresh_until: '2026-07-17T08:00:15.123+00:00',
    source_status: 'online',
    identity: {
      display_name: 'Lamball Keeper',
      account_name: 'keeper',
      level: 42,
    },
    map: {
      x: 123.5,
      y: -456.25,
      projection: 'palworld-world-v1',
      tile_set: 'default',
      tile_url: '/api/v1/map/tiles/{z}/{x}/{y}.png',
    },
    fields,
  }
}

describe('parsePresentation', () => {
  it('parses every available field kind as a discriminated union', () => {
    const parsed = parsePresentation(presentation([
      field({ id: 'identity.account', kind: 'text', value: 'keeper' }),
      field({ id: 'identity.level', kind: 'integer', value: 42 }),
      field({ id: 'activity.today', kind: 'duration_ms', value: 7_200_000 }),
      field({ id: 'presence.last_online', kind: 'timestamp', value: '2026-07-17T15:30:00+08:00' }),
      field({ id: 'network.latency', kind: 'latency_ms', value: 38.5 }),
      field({ id: 'location.coordinates', kind: 'coordinates', value: { x: 123.5, y: -456.25 } }),
      field({ id: 'presence.status', kind: 'status', value: '在线', tone: 'warning' }),
    ]))

    expect(parsed.fields.map(({ kind }) => kind)).toEqual([
      'text',
      'integer',
      'duration_ms',
      'timestamp',
      'latency_ms',
      'coordinates',
      'status',
    ])
    const coordinates = parsed.fields[5]
    if (coordinates.available && coordinates.kind === 'coordinates') {
      expectTypeOf(coordinates.value).toEqualTypeOf<{ x: number; y: number }>()
      expect(coordinates.value).toEqual({ x: 123.5, y: -456.25 })
    } else {
      throw new Error('expected available coordinates')
    }
    expectTypeOf(parsed.fields).toEqualTypeOf<DisplayField[]>()
    expectTypeOf<AvailableDisplayField>().not.toBeAny()
  })

  it('parses unavailable catalog entries without value or progress', () => {
    const rawField = field({
      id: 'network.latency',
      kind: 'latency_ms',
      available: false,
    })
    delete rawField.value
    const parsed = parsePresentation(presentation([
      rawField,
    ]))

    expect(parsed.fields[0]).toEqual({
      id: 'network.latency',
      label: '账号',
      kind: 'latency_ms',
      available: false,
      tone: 'normal',
    })
  })

  it('returns detached nested data without changing the input', () => {
    const input = presentation([
      field({ id: 'location.coordinates', kind: 'coordinates', value: { x: 1, y: 2 } }),
    ])
    const original = structuredClone(input)
    const parsed = parsePresentation(input)

    parsed.identity.display_name = 'Changed'
    parsed.fields[0].label = 'Changed'
    if (parsed.map) parsed.map.x = 0
    if (parsed.fields[0].available && parsed.fields[0].kind === 'coordinates') {
      parsed.fields[0].value.x = 0
    }

    expect(input).toEqual(original)
  })

  it('accepts every finite duration and latency number allowed by the protocol', () => {
    const parsed = parsePresentation(presentation([
      field({ id: 'activity.today', kind: 'duration_ms', value: 1.5 }),
      field({
        id: 'activity.week',
        kind: 'duration_ms',
        value: Number.MAX_SAFE_INTEGER + 1,
      }),
      field({ id: 'network.latency', kind: 'latency_ms', value: -0.5 }),
    ]))

    expect(parsed.fields.map((displayField) => {
      if (!displayField.available) throw new Error('expected available field')
      return displayField.value
    })).toEqual([1.5, Number.MAX_SAFE_INTEGER + 1, -0.5])
  })

  it('accepts a finite mathematical integer outside the safe-integer range', () => {
    const value = Number.MAX_SAFE_INTEGER + 1
    const parsed = parsePresentation(presentation([
      field({ id: 'identity.level', kind: 'integer', value }),
    ]))

    const parsedField = parsed.fields[0]
    if (!parsedField.available || parsedField.kind !== 'integer') {
      throw new Error('expected available integer field')
    }
    expect(parsedField.value).toBe(value)
  })

  it.each([
    ['a duplicate ID', [field(), field()]],
    ['an uppercase ID', [field({ id: 'Network.Latency' })]],
    ['an ID starting with punctuation', [field({ id: '.network' })]],
    ['an ID containing whitespace', [field({ id: 'network latency' })]],
    ['an overlong ID', [field({ id: `a${'b'.repeat(96)}` })]],
  ])('rejects %s', (_name, fields) => {
    expect(() => parsePresentation(presentation(fields))).toThrow()
  })

  it.each([
    ['impossible date', '2026-02-30T12:00:00Z'],
    ['missing zone', '2026-07-17T12:00:00'],
    ['invalid offset', '2026-07-17T12:00:00+24:00'],
    ['non-RFC3339 text', 'tomorrow'],
  ])('rejects %s timestamps', (_name, value) => {
    expect(() => parsePresentation(presentation([
      field({ id: 'presence.last_online', kind: 'timestamp', value }),
    ]))).toThrow(/RFC3339/)
  })

  it.each([
    ['integer fraction', 'integer', 1.5],
    ['NaN integer', 'integer', Number.NaN],
    ['infinite integer', 'integer', Number.POSITIVE_INFINITY],
    ['NaN duration', 'duration_ms', Number.NaN],
    ['infinite latency', 'latency_ms', Number.POSITIVE_INFINITY],
    ['infinite coordinate', 'coordinates', { x: Number.NEGATIVE_INFINITY, y: 0 }],
  ])('rejects %s', (_name, kind, value) => {
    expect(() => parsePresentation(presentation([
      field({ id: 'metric.value', kind, value }),
    ]))).toThrow()
  })

  it.each([
    ['text with number', 'text', 1],
    ['integer with string', 'integer', '1'],
    ['duration with string', 'duration_ms', '1000'],
    ['timestamp with number', 'timestamp', 1],
    ['latency with string', 'latency_ms', '38'],
    ['coordinates with array', 'coordinates', [1, 2]],
    ['coordinates with extra key', 'coordinates', { x: 1, y: 2, z: 3 }],
    ['status with boolean', 'status', true],
  ])('rejects %s', (_name, kind, value) => {
    expect(() => parsePresentation(presentation([
      field({ id: 'metric.value', kind, value }),
    ]))).toThrow()
  })

  it.each([
    ['unsupported kind', { kind: 'percentage' }],
    ['unsupported tone', { tone: 'urgent' }],
    ['progress below zero', { progress: -0.01 }],
    ['progress above one', { progress: 1.01 }],
    ['non-finite progress', { progress: Number.NaN }],
    ['non-number progress', { progress: '0.5' }],
  ])('rejects %s', (_name, overrides) => {
    expect(() => parsePresentation(presentation([field(overrides)]))).toThrow()
  })

  it.each([
    ['value', { available: false, value: 'stale' }],
    ['progress', { available: false, progress: 0.5 }],
  ])('rejects an unavailable field carrying %s', (_name, overrides) => {
    const rawField = field(overrides)
    if ('progress' in overrides) delete rawField.value
    expect(() => parsePresentation(presentation([rawField]))).toThrow()
  })

  it.each([
    ['an extra top-level key', { ...presentation(), future: true }],
    ['a missing top-level key', (() => { const value = presentation(); delete value.fields; return value })()],
    ['an extra field key', presentation([field({ future: true })])],
    ['a missing field key', presentation([(() => { const value = field(); delete value.label; return value })()])],
    ['a non-object field', { ...presentation(), fields: [null] }],
  ])('rejects %s', (_name, value) => {
    expect(() => parsePresentation(value)).toThrow()
  })
})
